src/main/kotlin/org/rust/ide/inspections/lints/RsUnnecessaryQualificationsInspection.kt

```diff
@@ -0,0 +1,87 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.inspections.lints
+
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiElement
+import com.intellij.psi.util.parentOfType
+import org.rust.ide.inspections.RsProblemsHolder
+import org.rust.ide.inspections.fixes.RemovePathPrefixFix
+import org.rust.lang.core.parser.RustParserUtil
+import org.rust.lang.core.psi.RsPath
+import org.rust.lang.core.psi.RsPathCodeFragment
+import org.rust.lang.core.psi.RsUseItem
+import org.rust.lang.core.psi.RsVisitor
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve.TYPES_N_VALUES_N_MACROS
+
+class RsUnnecessaryQualificationsInspection : RsLintInspection() {
+    override fun getLint(element: PsiElement): RsLint = RsLint.UnusedQualifications
+
+    override fun buildVisitor(holder: RsProblemsHolder, isOnTheFly: Boolean): RsVisitor = object : RsVisitor() {
+        override fun visitPath(path: RsPath) {
+            val root = path.rootPath()
+            val useItem = path.parentOfType<RsUseItem>()
+
+            if (useItem == null && root == path && root.canBeShortened()) {
+                val target = getUnnecessarilyQualifiedPath(path)
+                if (target != null) {
+                    val unnecessaryPart = target.fullQualifier()
+                    val range = TextRange(0, unnecessaryPart.length)
+
+                    val fix = RemovePathPrefixFix(
+                        path.containingFile,
+                        TextRange(root.startOffset, root.startOffset + unnecessaryPart.length)
+                    )
+                    holder.registerLintProblem(root, "Unnecessary qualification", range,
+                        RsLintHighlightingType.UNUSED_SYMBOL, listOf(fix))
+                }
+            }
+            super.visitPath(path)
+        }
+    }
+
+    private fun getUnnecessarilyQualifiedPath(path: RsPath): RsPath? {
+        if (path.resolveStatus == PathResolveStatus.UNRESOLVED) return null
+
+        val target = path.reference?.resolve() ?: return null
+
+        var pathText = path.referenceName ?: return null
+        var currentPath = path
+        val basePath = path.basePath()
+
+        while (true) {
```

Comment:
I don't like a bit that we use cycle with mutable variables here. Maybe we can analyze not only root paths in visitor, so we will not need to do manual recursion here?

Reply:
But then the inspection could be triggered multiple times? For
`a::b::c` I'm walking the path from the right, and once I find some suffix (`c`, `b::c`, `a::b::c`) that can be resolved, I mark everything to the left as unnecessary. If I analyzed `a::b` and `a::b::c` separately, I could mark the `a::` path as unnecessary twice if both `b::c` and `c` were resolvable.

I'm also not sure if we could depend on the visitor on visiting the path in the right order. I can try to refactor the while loop to a more functional implementation though.

Reply:
> But then the inspection could be triggered multiple times?

Yes, you are right

> I can try to refactor the while loop to a more functional implementation though.

Let's try. I would use something like this:
```kotlin
val qualifiers = generateSequence(path) { path.qualifier }
for (currentPath in qualifiers) {
    val currentName = currentPath.referenceName ?: break
    val pathText = "$currentName::${path.text.removePrefix(currentPath.text)}"
    ...
}
```